File: contests/vp/EC-Final 2025/A.cpp
#include<bits/stdc++.h>
using namespace std;
#define int long long
typedef __int128 i128;
struct Pt{
    i128 x,y;
};
Pt sub(Pt a,Pt b){return{a.x-b.x,a.y-b.y};}
i128 cross(Pt a,Pt b){return a.x*b.y-a.y*b.x;}
i128 dot(Pt a,Pt b){return a.x*b.x+a.y*b.y;}
signed main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    auto solve=[&](){
        int xa,ya,xb,yb,xc,yc;
        cin>>xa>>ya>>xb>>yb>>xc>>yc;
        int xd1,xd2,yd1,yd2;char s;
        cin>>xd1>>s>>xd2>>yd1>>s>>yd2;
        Pt A0={xa,ya},B0={xb,yb},C0={xc,yc};
        Pt AB0=sub(B0,A0),AC0=sub(C0,A0),BC0=sub(C0,B0);
        i128 L=(i128)xd2*yd2;
        Pt As={(i128)xa*L,(i128)ya*L},Bs={(i128)xb*L,(i128)yb*L},Cs={(i128)xc*L,(i128)yc*L};
        Pt Ds={(i128)xd1*yd2,(i128)yd1*xd2};
        Pt ABs=sub(Bs,As),ACs=sub(Cs,As),BCs=sub(Cs,Bs),ADs=sub(Ds,As);
        i128 hc=cross(ABs,ACs),hd=cross(ABs,ADs);
        if(hd==0||hd==hc){cout<<"Yes\n";return;}
        bool in_strip=(hc>0&&hd>0&&hd<=hc)||(hc<0&&hd<0&&hd>=hc);
        if(!in_strip){cout<<"No\n";return;}
        auto check=[&](Pt start_s,Pt dir_0,Pt base_0,Pt target_s)->bool{
            i128 num=cross(sub(target_s,start_s),dir_0);
            i128 den=cross(base_0,dir_0);
            if(den==0)return false;
            i128 modulus=den*L;
            return num%modulus==0;
        };
        if(check(As,AC0,AB0,Ds)||check(Bs,BC0,AB0,Ds)){cout<<"Yes\n";return;}
        if(dot(AB0,AC0)==0){
            i128 val=cross(sub(Ds,As),AC0);
            i128 limit=cross(sub(Bs,As),AC0);
            bool strictly_inside=false;
            if(limit>0)strictly_inside=(val>0&&val<limit);
            else if(limit<0)strictly_inside=(val<0&&val>limit);
            if(!strictly_inside){
                i128 da=dot(AC0,AC0),db=dot(BC0,AC0);
                Pt dir0;
                dir0.x=2*db*AC0.x-da*BC0.x;
                dir0.y=2*db*AC0.y-da*BC0.y;
                if(check(Cs,dir0,AB0,Ds)){cout<<"Yes\n";return;}
            }
        }
        if(dot(AB0,BC0)==0){
            i128 val=cross(sub(Ds,Bs),BC0);
            i128 limit=cross(sub(As,Bs),BC0);
            bool strictly_inside=false;
            if(limit>0)strictly_inside=(val>0&&val<limit);
            else if(limit<0)strictly_inside=(val<0&&val>limit);
            if(!strictly_inside){
                i128 db=dot(BC0,BC0),da=dot(AC0,BC0);
                Pt dir0;
                dir0.x=2*da*BC0.x-db*AC0.x;
                dir0.y=2*da*BC0.y-db*AC0.y;
                if(check(Cs,dir0,AB0,Ds)){cout<<"Yes\n";return;}
            }
        }
        cout<<"No\n";
    };
    int t;
    if(cin>>t)while(t--)solve();
    return 0;
}

File: contests/vp/EC-Final 2025/A.py
import sys
import math

def sub(a, b):
    return (a[0] - b[0], a[1] - b[1])

def cross(a, b):
    return a[0] * b[1] - a[1] * b[0]

def dot(a, b):
    return a[0] * b[0] + a[1] * b[1]

def check(start, dir_vec, base, target):
    num = cross(sub(target, start), dir_vec)
    den = cross(base, dir_vec)
    if den == 0:
        return False
    return num % den == 0

def parse_frac(s):
    if '/' in s:
        parts = s.split('/')
        return int(parts[0]), int(parts[1])
    return int(s), 1

def main():
    input_data = sys.stdin.read().strip().split()
    if not input_data:
        return
    
    t = int(input_data[0])
    idx = 1
    
    for _ in range(t):
        if idx + 7 >= len(input_data):
            break
            
        xa, ya, xb, yb, xc, yc = map(int, input_data[idx:idx+6])
        idx += 6
        
        xd_str = input_data[idx]
        yd_str = input_data[idx+1]
        idx += 2
        
        xd1, xd2 = parse_frac(xd_str)
        yd1, yd2 = parse_frac(yd_str)
        
        L = xd2 * yd2
        a = (xa * L, ya * L)
        b = (xb * L, yb * L)
        c = (xc * L, yc * L)
        d = (xd1 * yd2, yd1 * xd2)

        ab = sub(b, a)
        ac = sub(c, a)
        bc = sub(c, b)
        ad = sub(d, a)
        bd = sub(d, b)

        hc = cross(ab, ac)
        hd = cross(ab, ad)

        if hd == 0 or hd == hc:
            print("Yes")
            continue
        
        if not ((hc > 0 and 0 <= hd <= hc) or (hc < 0 and hc <= hd <= 0)):
            print("No")
            continue

        if hd == 0 or hd == hc or check(a, ac, ab, d) or check(b, bc, ab, d):
            print("Yes")
        elif dot(ab, ac) == 0:
            da = dot(ac, ac)
            db = dot(bc, ac)
            dir_vec = (2 * db * ac[0] - da * bc[0], 2 * db * ac[1] - da * bc[1])
            if check(c, dir_vec, ab, d):
                print("Yes")
            else:
                print("No")
        elif dot(ab, bc) == 0:
            db = dot(bc, bc)
            da = dot(ac, bc)
            dir_vec = (2 * da * bc[0] - db * ac[0], 2 * da * bc[1] - db * ac[1])
            if check(c, dir_vec, ab, d):
                print("Yes")
            else:
                print("No")
        else:
            print("No")

if __name__ == "__main__":
    main()